pkg/cmd/pr/shared/finder.go

```diff
@@ -321,70 +318,124 @@ func (f *finder) parseURL(prURL string) (ghrepo.Interface, int, error) {
-				prRefs.BranchName = strings.TrimPrefix(parsedPushRevision, r.Name+"/")
-				prRefs.HeadRepo = r.Repo
-				return prRefs, nil
-			}
+	// and branch name. It is of the form <remote>/<branch>. We suppress the error here because
+	// we have other means of computing the PullRequestRefs when this fails.
+	if pushRevisionRef, err := gitClient.PushRevision(context.Background(), localBranchName); err == nil {
+		remote, err := remotes.FindByName(pushRevisionRef.Remote)
+		if err != nil {
+			return PullRequestRefs{}, err
 		}
+		return PullRequestRefs{
+			BaseRepo:   baseRepo,
+			HeadRepo:   remote.Repo,
+			BranchName: pushRevisionRef.Branch,
+		}, nil
+	}
 
-		remoteNames := make([]string, len(rems))
-		for i, r := range rems {
-			remoteNames[i] = r.Name
-		}
-		return PullRequestRefs{}, fmt.Errorf("no remote for %q found in %q", parsedPushRevision, strings.Join(remoteNames, ", "))
+	// Otherwise, we'll look at branch and remote configuration to make a determination.
+	// We start by assuming that BaseRepo and HeadRepo are the same, and the branch name is
+	// the same as the local branch name, unless we find otherwise.
+	branchConfig, err := gitClient.ReadBranchConfig(context.Background(), localBranchName)
+	if err != nil {
+		return PullRequestRefs{}, err
 	}
 
-	// We assume the PR's branch name is the same as whatever f.BranchFn() returned earlier
-	// unless the user has specified push.default = upstream or tracking, then we use the
-	// branch name from the merge ref.
-	prRefs.BranchName = currentBranchName
-	if pushDefault == "upstream" || pushDefault == "tracking" {
-		prRefs.BranchName = strings.TrimPrefix(branchConfig.MergeRef, "refs/heads/")
+	pushDefault, err := gitClient.PushDefault(context.Background())
+	if err != nil {
+		return PullRequestRefs{}, err
+	}
+
+	// We assume the PR's branch name is the same as whatever was provided, unless the user has specified
+	// push.default = upstream or tracking, then we use the branch name from the merge ref.
+	remoteBranch := localBranchName
+	if pushDefault == git.PushDefaultUpstream || pushDefault == git.PushDefaultTracking {
+		remoteBranch = strings.TrimPrefix(branchConfig.MergeRef, "refs/heads/")
 	}
 
 	// To get the HeadRepo, we look to the git config. The HeadRepo comes from one of the following, in order of precedence:
 	// 1. branch.<name>.pushRemote
 	// 2. remote.pushDefault
 	// 3. branch.<name>.remote
 	if branchConfig.PushRemoteName != "" {
-		if r, err := rems.FindByName(branchConfig.PushRemoteName); err == nil {
-			prRefs.HeadRepo = r.Repo
-		}
-	} else if branchConfig.PushRemoteURL != nil {
-		if r, err := ghrepo.FromURL(branchConfig.PushRemoteURL); err == nil {
-			prRefs.HeadRepo = r
-		}
-	} else if remotePushDefault != "" {
-		if r, err := rems.FindByName(remotePushDefault); err == nil {
-			prRefs.HeadRepo = r.Repo
+		r, err := remotes.FindByName(branchConfig.PushRemoteName)
+		if err != nil {
+			// TODO: make error include remotes
```

Comment:
There's a few TODOs around better error messages that I'll drive out with the tests.